### PR TITLE
Some FSI additions

### LINKED ISSUE #23680: Make vector and field variables compatible across an interface when using interface kernels
## Reason
In some cases, you might have a vector variable on one side of an interface which interacts with, say three separate field variables representing x, y, and z component on the other. An example is in the FSI module: A number of the INS kernels require fluid velocity to be a vector variable; however, tensor mechanics kernels treat solid velocity, displacement, etc. as separate field variables such as disp_x, disp_y, etc. See discussion [#23658](https://github.com/idaholab/moose/discussions/23658).

## Design
I imagine something along the lines of: 
The user would specify a vector variable on one side, and, say 'neighbourVariable_x', 'neighbourVariable_y', etc. The number of neighbour components should match the dimension of the problem. For 'PenaltyCoupled' interface kernels, the 'primary_coupled_variable' would also be a vector, with corresponding 'secondary_coupled_variable_x', etc.
There might be a neater design than this, but this describes a basic functionality.

## Impact
This might require new interface kernels; however, it would be great if there's a way to update existing kernels to accept different variable types in this way and keep everything in one place!

### LINKED ISSUE #24462: Make ALE melt pool simulations run
## Reason
Users would like to be able to reproduce the ALE simulations in our MOOSE gallery

## Design
Update to new MOOSE objects and work on solver options

## Impact
Demonstrate capability for users

### PR DESCRIPTION
**Pull request by lindsayad**
- Interface kernel for coupling scalar field displacments to vector fluid velocity field
- Create ConvectedMeshPSPG so that we can have one-to-one comparison between INS and INSAD in the FSI module
- Fix mesh convection residual for INSAD. The multiplication order needs to be switched to match the regular advection term

Closes #23680

### CODE REVIEW

**Comment by lindsayad** [comment]
@eilloo I think the work here is relevant to some of your discussion posts as well as #23680

**Comment by lindsayad** [comment]
For ease of reviewing I'm going to split this into smaller chunks starting with #24514 and #24515

**Comment by lindsayad** [comment]
With #24515 merged, turning to #24540

**Comment by lindsayad** [comment]
Once the NS PR goes in then I will return here for the FSI code

**Review by maxnezdyur — COMMENTED** [review]
Looking at how MeshConvection was implemented for INSAD, everything looks correct, but I think the relative velocity would need to be used in place of the fluid velocity for a consistent stabilization. I don't know how important that is here but wanted to make you aware of it.

**Comment by lindsayad** [comment]
> but I think the relative velocity would need to be used in place of the fluid velocity for a consistent stabilization.

Looking at [this 2019 article](https://www.researchgate.net/profile/Tayfun-Tezduyar/publication/331353222_A_Stabilized_ALE_Method_for_Computational_Fluid-Structure_Interaction_Analysis_of_Passive_Morphing_in_Turbomachinery/links/5cfc6fe94585157d159e089a/A-Stabilized-ALE-Method-for-Computational-Fluid-Structure-Interaction-Analysis-of-Passive-Morphing-in-Turbomachinery.pdf) I do see the modification to the velocity used to multiply the test function gradient in the SUPG term but I do not see it used in tau. However, that seems like an oversight to me. The momentum advection will be defined by the relative velocity (e.g. u - u_hat) so it makes sense to me that if the relative velocity is 0, then the contribution to tau from advection should be 0

**Comment by maxnezdyur** [comment]
I thought the advection-dominated limit term would also take into account the mesh velocity, but I can't seem to find my source, so I may be mistaken there. Basically, I would expect everywhere we use `U.norm()` in our tau parameters we would use `(U-U_hat).norm()`. I'll try to find my source though.

**Comment by lindsayad** [comment]
Source would be great, but yea I agree that seems logical

**Comment by lindsayad** [comment]
This now sits on top of https://github.com/idaholab/moose/pull/25574 so probably should wait for further review until that is merged

**Review by maxnezdyur — APPROVED** [review]

**Comment by lindsayad** [comment]
@maxnezdyur has reviewed this. @loganharbour, @GiudGiud, or @roystgnr could I get a CCB review when one of you has a chance?

File: modules/navier_stokes/include/materials/INSADTauMaterial.h [hunk 1]
```diff
@@ -90,11 +90,15 @@ class INSADTauMaterialTempl : public T
   /// The velocity system number
   const unsigned int _vel_sys_number;
 
+  /// The speed of the medium. This is the norm of the relative velocity, e.g. the velocity minus
+  /// the mesh velocity
```
Comment (GiudGiud) [inline modules/navier_stokes/include/materials/INSADTauMaterial.h:94]:
```suggestion
  /// the mesh velocity, at the current _qp
```

File: modules/navier_stokes/include/kernels/INSBase.h [hunk 2]
```diff
@@ -106,4 +110,13 @@ class INSBase : public Kernel
   bool _laplace;
   bool _convective_term;
   bool _transient_term;
+
+  /// Whether displacements are provided
+  const bool _disps_provided;
+  /// Time derivative of the x-displacement
```
Comment (GiudGiud) [inline modules/navier_stokes/include/kernels/INSBase.h:116]:
```suggestion
  /// Time derivative of the x-displacement, mesh velocity in the x-direction
```

File: modules/fsi/test/tests/2d-small-strain-transient/tests [hunk 3]
```diff
@@ -1,14 +1,42 @@
...
+  issues = '#12462'
+  [fsi]
+    requirement = 'We shall be able to combine navier-stokes and tensor-mechanics simulation capabilities to do some basic fluid-structure interaction simulations and produce equivalent results when the Navier-Stokes equations are implemented using'
+    [INS]
+      type = 'Exodiff'
+      input = 'fsi_flat_channel.i'
+      exodiff = 'fsi_flat_channel_out.e'
+      detail = 'scalar field variables for velocity components and hand-coded Jacobians,'
+      # vel_x_solid is not accurate
+      # We mmay improve that in the future
```
Comment (GiudGiud) [inline modules/fsi/test/tests/2d-small-strain-transient/tests:12]:
```suggestion
      # We may improve that in the future
```

File: modules/fsi/include/kernels/ConvectedMeshPSPG.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "INSBase.h"
+
+/**
+ * This calculates the time derivative for a coupled variable
```
Comment (GiudGiud) [inline modules/fsi/include/kernels/ConvectedMeshPSPG.h:15]:
```suggestion
 *  Computes residual and Jacobian contributions for the PSPG stabilization term for mesh advection
```

File: modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    const auto & dof_indices = _velocity_var->dofIndices();
+    mooseAssert(phi.size() == dof_indices.size(), "These should be the same");
+
+    _residuals.resize(phi.size());
+    for (auto & r : _residuals)
+      r = 0;
+
+    for (const auto qp : make_range(_qrule->n_points()))
+      for (const auto i : index_range(phi))
+        _residuals[i] += phi[i][qp] * _qp_jumps[qp];
```
Comment (GiudGiud) [inline modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C:82]: I would expect flipping the loops to be faster

File: modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+void
+ADPenaltyVelocityContinuity::computeNeighborOffDiagJacobian(unsigned int)
+{
+}
+
+const MooseVariableFieldBase &
+ADPenaltyVelocityContinuity::variable() const
+{
+  return *_velocity_var;
+}
```
Comment (GiudGiud) [inline modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C:139]: header?

File: modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C (hunk 5 above)
Comment (lindsayad) [reply modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C:82]: why?

Comment (GiudGiud) [reply modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C:82]: more local? you re jumping around in memory for phi array every i (i * qp times) here.

the other one you loop on the inner index and you only jump in memory for phi i times

Comment (GiudGiud) [reply modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C:82]: probably doesnt matter.

Comment (lindsayad) [reply modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C]: change

**Review by GiudGiud — APPROVED** [review]

**Comment by lindsayad** [comment]
thanks for the reviews!

**Pull request merged**